Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs

```diff
@@ -0,0 +1,78 @@
+﻿// © 2021 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Maintainability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidAsyncVoidAnalyzer : DiagnosticAnalyzer
+	{
+		private const string Title = @"Avoid async void";
+		public const string MessageFormat = @"Methods may not have async void return type";
+		private const string Description = @"To avoid unhandled exception, methods should not use async void unless a event handler.";
+		private const string Category = Categories.Maintainability;
+		private const string HelpUri = @"https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void";
+		
+		private static DiagnosticDescriptor Rule = new DiagnosticDescriptor(Helper.ToDiagnosticId(DiagnosticIds.AvoidAsyncVoid), Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description, helpLinkUri: HelpUri);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterCompilationStartAction(startContext =>
+			{
+				if (startContext.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Task") == null)
+				{
+					return;
+				}
+				startContext.RegisterSyntaxNodeAction(Analyze, SyntaxKind.MethodDeclaration);
+			});
+		}
+
+		private void Analyze(SyntaxNodeAnalysisContext context)
+		{
+			MethodDeclarationSyntax methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+
+			if (!methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword))
```

Comment:
I think MethodDeclarationSyntax also has a ReturnType, for which you can compare with void.  Ie if not void, return.

Reply:
Yes. I added that check with a void return type. I think made it little complicated before. Now the logic is simple and better.